### PR TITLE
First service in backend

### PR DESCRIPTION
**Pull request by nyxordinal**
* update backend's README
* update models and schemas
* create endpoint for get all messages and create message

### CODE REVIEW

File: packages/backend/src/controllers/message.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { getAllMessage, storeMessage } from '@service/message'
+import {
+  responseBadRequest,
+  responseCreated,
+  responseInternalServerError,
+  responseSuccess,
+} from '@util/response'
+import { Request, Response } from 'express'
+
+export const getAllMessageController = async (_: Request, res: Response) => {
```
Comment (azuraga) [inline packages/backend/src/controllers/message.ts]: maybe change to getAllMessage**s**Controller and update related entries accordingly

File: packages/backend/src/services/message.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { MessageDoc, MessageInterface, MessageModel } from '@model/message'
+
+/**
+ * Get all message
```
Comment (azuraga) [inline packages/backend/src/services/message.ts]:
```suggestion
 * Get all messages
```

File: packages/backend/src/controllers/message.ts (hunk 1 above)
Comment (azuraga) [reply packages/backend/src/controllers/message.ts]: PS: deliberately didn't add this as a suggestion since references should also be updated

**Comment by PS-Soundwave** [comment]
So just a general question: Isn't separating verified and unverified records into separate collections a good idea, since essentially every query will be conditioned on verification?

**Comment by RedMaps** [comment]
> Isn't separating verified and unverified records into separate collections a good idea, since essentially every query will be conditioned on verification?

Can't we also just index that query instead?

**Comment by nyxordinal** [comment]
Yes, we can create index based on `is_verified` properties

**Comment by RedMaps** [comment]
Sounds like the better option of the two to me then

**Comment by nyxordinal** [comment]
Since this is the first service, It's still very simple and I made this specifically for future service reference 
After this PR got merged, I'll try to create issues for other services and some improvement
So others can contribute to the backend

File: README.md [hunk 3]
```diff
@@ -33,9 +33,62 @@ The `globals.scss` is mainly used for standardization and should rarely be touch
...
 
-> The backend structure is still up to debate and will change in the near future.
-
-The backend uses express as a server for hosting a REST API, which is all set up inside the `index.ts`.
+The backend uses [Express.js](https://expressjs.com/) as server for hosting a REST API, which is all set up inside the `src/index.ts`.
+Backend also uses several other library:
+
+- [Mongoose](https://expressjs.com/) as ODM for Mongo DB
+- [cors](https://www.npmjs.com/package/cors) as enabler for CORS request
+- [morgan](https://www.npmjs.com/package/morgan) and [winston](https://www.npmjs.com/package/winston) for log any activity that happen inside backend service
```
Comment (RedMaps) [inline README.md]:
```suggestion
Backend also uses several other library:

- [Mongoose](https://expressjs.com/) as ODM for Mongo DB
- [cors](https://www.npmjs.com/package/cors) as enabler for CORS requests
- [morgan](https://www.npmjs.com/package/morgan) and [winston](https://www.npmjs.com/package/winston) for logging any activity that happens inside the backend service
```

File: README.md [hunk 4]
```diff
@@ -33,9 +33,62 @@ The `globals.scss` is mainly used for standardization and should rarely be touch
...
+- `index.ts` => entry point to backend
+- `logger.ts` => file for logger
+- `route.ts` => place all routes here
+- `db.ts` => file for MongoDB connection
+- `config` => store any shared config across backend here
+- `controllers` => place where you write handlers for route
+- `middleware` => folder for middlewares
+- `models` => folder for MongoDB models
+- `schemas` => folder for MongoDB schemas
+- `services` => where you store any logic code related to data processing before it is saved or fetched to DB
```
Comment (RedMaps) [inline README.md]:
```suggestion
- `src` => stores all backend files
- `index.ts` => entry point to backend
- `logger.ts` => file for logger
- `route.ts` => place all routes here
- `db.ts` => file for MongoDB connection
- `config` => store any shared config across backend here
- `controllers` => place where you write handlers for routes
- `middleware` => folder for middlewares
- `models` => folder for MongoDB models
- `schemas` => folder for MongoDB schemas
- `services` => where you store any logic code related to data processing before it is saved or fetched to DB
```

**Comment by PS-Soundwave** [comment]
I'm also moderately concerned about getAllMessages for traffic reasons but if it's improved before deployment I don't suppose it matters.

File: README.md [hunk 5]
```diff
@@ -33,9 +33,62 @@ The `globals.scss` is mainly used for standardization and should rarely be touch
...
+- `utils` => folder for util functions
+
+#### Logs
+
+After you start the backend server, a folder will be created automatically, namely the `logs` folder. There are 5 log files that will be created and updated as the server runs, here is the actual place where any activity logs on the backend will be recorded.
```
Comment (RedMaps) [inline README.md]:
```suggestion
After you start the backend server, a folder will be created automatically, named the `logs` folder. There are 5 log files that will be created and updated as the server runs, here is the actual place where any activity logs on the backend will be recorded.
```

**Comment by azuraga** [comment]
> I'm also moderately concerned about getAllMessages for traffic reasons but if it's improved before deployment I don't suppose it matters.

Ah yes this definitely could e.g. some paging but that's indeed for another issue imo

File: README.md [hunk 6]
```diff
@@ -33,9 +33,62 @@ The `globals.scss` is mainly used for standardization and should rarely be touch
...
+
+#### ENV Variables
+
+Supposedly, env variable's name inside `package/backend/.env.example` is self explanatory. Here is more explanation of them:
+
+- `PORT` => Backend service port
+- `DEBUG` => Set to `TRUE` if you want debug messages to show, ideally in dev you set this to `TRUE` and `FALSE` in production
+- `CLIENT_URL` => frontend's URL
+- `DB_URI` => mongodb connection string, [you can find the detail of it here](https://docs.mongodb.com/manual/reference/connection-string/)
```
Comment (RedMaps) [inline README.md]:
```suggestion
Env variable names inside `package/backend/.env.example` should be self explanatory. Here is more explanation of them:

- `PORT` => Backend service port
- `DEBUG` => Set to `TRUE` if you want debug messages to show, ideally in dev you set this to `TRUE` and to `FALSE` in production
- `CLIENT_URL` => frontend URL
- `DB_URI` => mongodb connection string, [you can find the details for it here](https://docs.mongodb.com/manual/reference/connection-string/)
```

**Comment by nyxordinal** [comment]
> Ah yes this definitely could e.g. some paging but that's indeed for another issue imo

Yep, I'll definitely add pagination in one of the issues

File: packages/backend/src/controllers/front.ts [hunk 7]
```diff
@@ -4,5 +4,10 @@ import { responseSuccess } from '@util/response'
 
 export const indexController = (req: Request, res: Response) => {
   Logger.info('Enter index route')
-  return responseSuccess(res, null, 'Welcome to Kaichou Project API')
+
+  const data = {
+    status: 'online',
+    time: new Date(),
+  }
+  return responseSuccess(res, data, 'Welcome to Kaichou Project API')
```
Comment (RedMaps) [inline packages/backend/src/controllers/front.ts]:
```suggestion
  return responseSuccess(res, data, 'Welcome to the Kaichou Project API')
```

File: packages/backend/src/models/message.ts [hunk 8]
```diff
@@ -3,9 +3,10 @@ import { MODEL_MESSAGE } from '@constant/general'
 import { MessageSchema } from '@model/schemas/message'
 
 export interface MessageInterface {
-  author: string
-  avatarUrl: string
+  creator: string
+  avatarUrl: string //TODO: utilize this if needed, if not just remove it from model
```
Comment (RedMaps) [inline packages/backend/src/models/message.ts]: Yeah i had problems integrating this with the designs... I had considered instead giving people a few choices of decoration elements for the top right of the card. (This is nothing to change, I just put this here for future reference)

**Review by RedMaps — CHANGES_REQUESTED** [review]
This is about all the review i can do from my phone right now. Maybe someone else can jump in on actually pulling the pr and testing ^^ Mainly just grammar and some misc comments.

File: packages/backend/src/models/message.ts (hunk 8 above)
Comment (nyxordinal) [reply packages/backend/src/models/message.ts]: So is it better to remove the `avatarUrl` or just keep it for now?

**Review by RedMaps — APPROVED** [review]
Seems good now!

File: packages/backend/src/models/message.ts (hunk 8 above)
Comment (RedMaps) [reply packages/backend/src/models/message.ts]: Hmm i'd like to hear the opinions of the others on this

Comment (RedMaps) [reply packages/backend/src/models/message.ts]: So we decided that keeping this doesn't make sense right now. If I decide to design decorations ill add the field for those myself or at least open an issue so dont bother with that right now.

**Pull request merged**